File: dclib/core/src/main/java/dclib/epf/graphics/EntityTransformDrawer.java
package dclib.epf.graphics;

import java.util.List;

import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer.ShapeType;

import dclib.epf.Entity;
import dclib.epf.parts.TransformPart;
import dclib.graphics.ScreenHelper;
import dclib.physics.DefaultTransform;

public final class EntityTransformDrawer implements EntityDrawer {

	private final ShapeRenderer shapeRenderer;
	private final ScreenHelper screenHelper;

	public EntityTransformDrawer(final ShapeRenderer shapeRenderer, final ScreenHelper screenHelper) {
		this.shapeRenderer = shapeRenderer;
		this.screenHelper = screenHelper;
	}

	@Override
	public final void draw(final List<Entity> entities) {
		screenHelper.setScaledProjectionMatrix(shapeRenderer);
		shapeRenderer.begin(ShapeType.Line);
		for (Entity entity : entities) {
			draw(entity);
		}
		shapeRenderer.end();
	}

	private void draw(final Entity entity) {
		TransformPart transformPart = entity.tryGet(TransformPart.class);
		if (transformPart != null) {
			if (transformPart.getTransform() instanceof DefaultTransform) {
				DefaultTransform transform = (DefaultTransform)transformPart.getTransform();
				float[] transformedVertices = transform.getTransformedVertices();
				shapeRenderer.polygon(transformedVertices);
			}
		}
	}

}

File: dclib/core/src/main/java/dclib/epf/graphics/EntityTransformDrawer.kt
package dclib.epf.graphics

import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.badlogic.gdx.graphics.glutils.ShapeRenderer.ShapeType
import dclib.epf.Entity
import dclib.epf.parts.TransformPart
import dclib.graphics.ScreenHelper
import dclib.physics.DefaultTransform

class EntityTransformDrawer(
        private val shapeRenderer: ShapeRenderer,
        private val screenHelper: ScreenHelper)
    : EntityDrawer {
    override fun getName(): String {
        return "transform"
    }

    override fun draw(entities: Collection<Entity>) {
        screenHelper.setScaledProjectionMatrix(shapeRenderer)
        shapeRenderer.begin(ShapeType.Line)
        for (entity in entities) {
            draw(entity)
        }
        shapeRenderer.end()
    }

    private fun draw(entity: Entity) {
        val transformPart = entity.tryGet(TransformPart::class)
        if (transformPart != null) {
            val transform = transformPart.transform
            if (transform is DefaultTransform) {
                shapeRenderer.polygon(transform.transformedVertices)
            }
        }
    }
}
